app/lib/encryption/encryption.js

```diff
@@ -0,0 +1,420 @@
+		const db = database.active;
+
+		const messagesCollection = db.collections.get('messages');
+		const threadsCollection = db.collections.get('threads');
+		const threadMessagesCollection = db.collections.get('thread_messages');
+
+		// e2e status is 'pending' and message type is 'e2e'
+		const whereClause = [
+			Q.where('t', E2E_MESSAGE_TYPE),
+			Q.where('e2e', E2E_STATUS.PENDING)
+		];
+
+		// decrypt messages of a room
+		if (roomId) {
+			whereClause.push(Q.where('rid', roomId));
+		}
+
+		try {
+			// Find all messages/threads/threadsMessages that have pending e2e status
+			const messagesToDecrypt = await messagesCollection.query(...whereClause).fetch();
+			const threadsToDecrypt = await threadsCollection.query(...whereClause).fetch();
+			const threadMessagesToDecrypt = await threadMessagesCollection.query(...whereClause).fetch();
+
+			// Concat messages/threads/threadMessages
+			let toDecrypt = [...messagesToDecrypt, ...threadsToDecrypt, ...threadMessagesToDecrypt];
+			toDecrypt = await Promise.all(toDecrypt.map(async(message) => {
+				const { rid, t, msg } = message;
+				const newMessage = await this.decryptMessage({ rid, t, msg });
+				if (message._hasPendingUpdate) {
+					console.log(message);
+					return;
+				}
+				return message.prepareUpdate(protectedFunction((m) => {
+					Object.assign(m, newMessage);
+				}));
+			}));
+
+			await db.action(async() => {
+				await db.batch(...toDecrypt);
+			});
+		} catch {
+			// Do nothing
+		}
+	}
+
+	// Logic to decrypt all pending subscriptions
+	// after initialize the encryption client
+	decryptPendingSubscriptions = async() => {
+		const db = database.active;
+		const subCollection = db.collections.get('subscriptions');
+		try {
+			// Find all rooms that can have a lastMessage encrypted
+			// If we select only encrypted rooms we can miss some room that changed their encrypted status
+			const subsEncrypted = await subCollection.query(Q.where('e2e_key_id', Q.notEq(null))).fetch();
+			// We can't do this on database level since lastMessage is not a database object
+			const subsToDecrypt = subsEncrypted.filter(sub => (
+				// Encrypted message
+				sub?.lastMessage?.t === E2E_MESSAGE_TYPE
+				// Message pending decrypt
+				&& sub?.lastMessage?.e2e === E2E_STATUS.PENDING
+			));
+			await Promise.all(subsToDecrypt.map(async(sub) => {
+				const { lastMessage } = sub;
+				const newSub = await this.decryptSubscription({ lastMessage });
+				if (sub._hasPendingUpdate) {
+					console.log(sub);
+					return;
+				}
+				return sub.prepareUpdate(protectedFunction((m) => {
+					Object.assign(m, newSub);
+				}));
+			}));
+
+			await db.action(async() => {
+				await db.batch(...subsToDecrypt);
+			});
+		} catch {
+			// Do nothing
+		}
+	}
+
+	// Decrypt a subscription lastMessage
+	decryptSubscription = async(subscription) => {
+		// If the subscription doesn't have a lastMessage just return
+		if (!subscription?.lastMessage) {
+			return subscription;
```

Comment:
Not possible, we're decrypting data received from servers, so, it should be updated always, since the field that should updated can be something different of the lastMessage, we should return the new object to be inserted/updated on our local database.